services/src/main/java/org/keycloak/services/resources/admin/permissions/RolePermissionsV2.java

```diff
@@ -0,0 +1,339 @@
+        }
+
+        return hasPermission(role, MAP_ROLE_CLIENT_SCOPE_SCOPE);
+    }
+
+    @Override
+    public boolean canMapComposite(RoleModel role) {
+        if (canManageDefault(role)) return checkAdminRoles(role);
+
+        if (role.getContainer() instanceof ClientModel) {
+            if (root.clients().canMapCompositeRoles((ClientModel) role.getContainer())) return true;
+        }
+
+        return hasPermission(role, MAP_ROLE_COMPOSITE_SCOPE) && checkAdminRoles(role);
+    }
+
+    @Override
+    public boolean canMapRole(RoleModel role) {
+        if (root.hasOneAdminRole(AdminRoles.MANAGE_USERS)) return checkAdminRoles(role);
+
+        if (role.getContainer() instanceof ClientModel) {
+            if (root.clients().canMapRoles((ClientModel) role.getContainer())) return true;
+        }
+
+        return hasPermission(role, MAP_ROLE_SCOPE) && checkAdminRoles(role);
+    }
+
+    @Override
+    public Set<String> getRolesWithPermission(String scope) {
+        if (!root.isAdminSameRealm()) {
+            return Collections.emptySet();
+        }
+
+        ResourceServer server = root.realmResourceServer();
+
+        if (server == null) {
+            return Collections.emptySet();
+        }
+
+        Set<String> granted = new HashSet<>();
+
+        resourceStore.findByType(server, AdminPermissionsSchema.ROLES_RESOURCE_TYPE, resource -> {
+            if (hasGrantedPermission(server, resource, scope)) {
+                granted.add(resource.getName());
+            }
+        });
+
+        return granted;
+    }
+
+    private boolean hasPermission(RoleModel role, String scope) {
+        if (!root.isAdminSameRealm()) {
+            return false;
+        }
+        ResourceServer server = root.realmResourceServer();
+        if (server == null) return false;
+
+        Resource resource =  resourceStore.findByName(server, role.getId(), server.getId());
+        if (resource == null) {
+            // check if there is permission for "all-roles". If so, load its resource and proceed with evaluation
+            resource = AdminPermissionsSchema.SCHEMA.getResourceTypeResource(session, server, AdminPermissionsSchema.ROLES_RESOURCE_TYPE);
+
+            if (authz.getStoreFactory().getPolicyStore().findByResource(server, resource).isEmpty()) {
+                return false;
+            }
+        }
+
+        return hasGrantedPermission(server, resource, scope);
+    }
+
+    private boolean hasGrantedPermission(ResourceServer server, Resource resource, String scope) {
+        Collection<Permission> permissions = root.evaluatePermission(new ResourcePermission(resource, resource.getScopes(), server), server);
+        for (Permission permission : permissions) {
+            for (String s : permission.getScopes()) {
+                if (scope.equals(s)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    // TODO do we need this terrible thing?
```

Comment:
Yes, we need. The reason is that this method is used to check whether the admin user has management roles and eventually grant access to the admin console.

I have worked on an issue recently that was going through this method and I think we can consider reviewing it later. But not now ...

Reply:
OK, I kept it there